### PR TITLE
feat: torch.compile encoder + AutoPrecisionCallback + compile benchmark

### PR DESCRIPTION
**Pull request by felixmaximilian**
Stacked on #231.

Port of PR #220 onto `feat/export-unification`. Core compile logic is unchanged from PR #220 to avoid requiring re-benchmarking.

## What this does

- **`ControlTransformer`**: add `compile: bool = False` and `compile_backend: str = "inductor"` params. When `compile=True`, wraps encoder with `torch.compile`.
- **`AttentionMaskLegend`**: `@tensorclass` → `NamedTuple`. Legend values become Python scalars that Dynamo inlines as constants — required for cudagraph capture (0-dim tensors caused cudagraph to treat them as dynamic inputs) and `torch.export` compatibility.
- **`AutoPrecisionCallback`**: downgrades `bf16-mixed` to `16-mixed` on GPUs without native bf16 support (e.g. Turing). Added to `pretrain.yaml` and `finetune.yaml` callback configs.
- **`test_compile_benchmark.py`**: wall-clock benchmark, eager vs compiled encoder. Auto-probes best backend (`inductor → cudagraphs → aot_eager`). `--backends` flag for explicit sweeps. Asserts compiled ≤ 1.5× slower than eager.
- **`justfile`**: add `PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True`.
- **`flake.nix`**: export `CXX/CC` to system gcc on Linux for inductor's C extension build.
- **`config/datamodule/yaak/train.yaml`**: add `drop_last: true`.
- **`config/datamodule/yaak/train_debug.yaml`**: new debug datamodule config.

## Why compile only the encoder

The encoder (transformer body) is the compute-heavy portion — it benefits most from kernel fusion and AOT tracing. The episode builder and objectives contain Python-level dispatch that isn't currently compile-friendly.

## Backend benchmark results (encoder forward only)

Measured with `--backends inductor,cudagraphs,aot_eager`:

| backend | CUDA speedup | CPU speedup |
|---|---|---|
| **inductor** | **1.26×** | **5.83×** |
| cudagraphs | 1.13× | 1.40× |
| aot_eager | 1.13× | 1.31× |

**`inductor` is the default.**

### RTX 5090 / Blackwell — bf16-mixed, bs=90, drop_last=true

| config | epoch 0 | steady state |
|---|---|---|
| no compile | ~0.71 it/s | ~0.71 it/s |
| **compile (inductor)** | 0.53 it/s (JIT warmup) | **0.83 it/s** |

**~17% throughput gain** in steady state. Epoch 0 penalty is one-time JIT compilation cost. Inductor wins here because Blackwell has native bf16 tensor cores for kernel fusion.

### TITAN RTX / Turing — bf16-mixed vs 16-mixed

| config | steady state |
|---|---|
| bf16-mixed, no compile | 0.10 it/s |
| **16-mixed, no compile** | **~0.20–0.29 it/s** |

The big win on Turing is precision (`AutoPrecisionCallback` downgrade), not compile — Turing has no native bf16 tensor cores.

## What had to be done differently vs PR #220

**`embeddings_unpacked` → `token_embeddings`** in `test_compile_benchmark.py` — #231 renamed this field.

**`AttentionMaskLegend.from_tensor` simplified** — no `torch.as_tensor` wrapping needed since values are now plain Python scalars stored in the NamedTuple.

**`_soft_pending_unbacked` fixture added** — PR #220 used `EpisodeExport` (a plain dataclass) as the export output. With #231's unified `Episode` (TensorClass) path, tensordict's `_device_recorder` and `_CONSTRUCTORS` globals are mutated during export tracing. Dynamo detects these as side effects and creates a spurious "pending unbacked symbol `u0`" error even though the exported graph is valid. Setting `soft_pending_unbacked_not_found_error = True` demotes it to a warning. The export is verified correct by `test_onnx_inference` (ORT numerics match eager).

## Test plan
- [x] 62 tests pass, 2 xfailed (`test_torch_export_dynamic_shapes`)

### CODE REVIEW

**Comment by felixmaximilian** [comment]
Ready for final review.

File: src/rmind/callbacks/precision.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+        if not isinstance(plugin, MixedPrecision):
+            return
+        if plugin.precision != "bf16-mixed":
+            return
+        if torch.cuda.is_bf16_supported():
+            return
+
+        plugin.precision = "16-mixed"
+        if plugin.scaler is None:
+            plugin.scaler = torch.cuda.amp.GradScaler()
```
Comment (egorchakov) [inline src/rmind/callbacks/precision.py]: this kind of lightning plugin surgery feels a little fragile -- perhaps either

a. swap the entire `trainer.precision_plugin`
b. validate trainer args pre-instantiation and error/warning out if `precision == "bf16-mixed"` but `torch.cuda.is_bf16_supported() == False`

File: src/rmind/models/control_transformer.py [hunk 2]
```diff
@@ -89,6 +91,9 @@ def __init__(  # noqa: PLR0913
             hparams["encoder"] = encoder.model_dump()
             encoder = encoder.instantiate()
 
+        if compile:
+            encoder = torch.compile(encoder, backend=compile_backend)  # ty:ignore[invalid-assignment]
+
```
Comment (egorchakov) [inline src/rmind/models/control_transformer.py]: consider keeping this config-only a la

```yaml
encoder:
  _target_: torch.compile
  # disable: false
  model: ...
```

File: src/rmind/models/control_transformer.py [hunk 3]
```diff
@@ -279,6 +284,41 @@ def forward(self, batch: TensorTree) -> TensorTree | TensorDict:
...
+        else:
+            # Model is not compiled: keys must be encoder.*
+            # Strip prefix from any legacy encoder._orig_mod.* keys.
+            checkpoint["state_dict"] = {
+                k.replace("encoder._orig_mod.", "encoder.", 1)
+                if k.startswith("encoder._orig_mod.")
+                else k: v
+                for k, v in state.items()
+            }
+
```
Comment (egorchakov) [inline src/rmind/models/control_transformer.py]: (as discussed) so it seems like it's either of:

- config-only `torch.compile` and a generalized version of `on_save/load_checkpoint` that handles all `_orig_mod` not just `encoder._orig_mod`
- config+constructor `torch.nn.Module.compile` (inplace) with no checkpoint surgery (https://github.com/Lightning-AI/pytorch-lightning/issues/20233#issuecomment-2868169706)

File: tests/test_compile_benchmark.py
Comment (egorchakov) [inline tests/test_compile_benchmark.py:1]: perhaps worth considering https://docs.pytorch.org/docs/2.12/user_guide/torch_compiler/torch.compiler_profiling_torch_compile.html 

(can probably just dump to stdout vs exporting chrome traces but TBD)

File: flake.nix [hunk 4]
```diff
@@ -29,6 +29,10 @@
               (lib.optionalString stdenv.isDarwin "export DYLD_FALLBACK_LIBRARY_PATH=${
                 pkgs.lib.makeLibraryPath [ pkgs.ffmpeg ]
               }")
+              (lib.optionalString stdenv.isLinux ''
+                export CXX=/usr/bin/g++
+                export CC=/usr/bin/gcc
```
Comment (egorchakov) [inline flake.nix]: 1. what do we need this for? AOTInductor benchmarks?
2. could we have those managed by nix rather than relying on system binary availability?

File: README.md [hunk 5]
```diff
@@ -42,6 +42,18 @@ Foundation models for spatial intelligence.
 just train experiment=yaak/control_transformer/pretrain [...]
 ```
 
+Training uses `torch.compile` on the encoder by default (set in the model config). To disable it, pass `compile=false`.
+
+### Debug training (3 episodes, no compile)
```
Comment (egorchakov) [inline README.md:47]: 🚀

**Review by egorchakov — CHANGES_REQUESTED** [review]

File: config/datamodule/yaak/train_debug.yaml [hunk 6]
```diff
@@ -0,0 +1,29 @@
+---
+defaults:
+  - /dataset/yaak/train_debug@train.dataset
+  - /dataset/yaak/val@val.dataset
```
Comment (koritsky) [inline config/datamodule/yaak/train_debug.yaml]: `dataset/yaak/val.yaml` is still loaded here. consider loading the same dataset config to val dataloader as well. probablyrename it to just `debug.yaml` since we dont actually train it. 

OR drop val dataloader at all, but afaik some sanity check callbacks use val dataloader

File: config/datamodule/yaak/train_debug.yaml [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+  - _self_
+
+_target_: rmind.datamodules.GenericDataModule
+train:
+  _target_: rbyte.dataloader.TorchDataNodeDataLoader
+  batch_size: 64
```
Comment (koritsky) [inline config/datamodule/yaak/train_debug.yaml]: reduce batch size?

File: config/experiment/yaak/control_transformer/pretrain_debug.yaml [hunk 8]
```diff
@@ -0,0 +1,8 @@
+# @package _global_
+
+defaults:
+  - pretrain
+  - /datamodule: yaak/train_debug
```
Comment (koritsky) [inline config/experiment/yaak/control_transformer/pretrain_debug.yaml]: prob not worth  having an experiment config for that. i believe it can be just provided as a cli hydra arg and/or a line in justfule for debug setup

File: config/trainer/callbacks/finetune.yaml [hunk 9]
```diff
@@ -1,3 +1,4 @@
+- _target_: rmind.callbacks.AutoPrecisionCallback
```
Comment (koritsky) [inline config/trainer/callbacks/finetune.yaml]: in callbacks + yamls we trust
🥇

**Review by koritsky — DISMISSED** [review]

File: src/rmind/models/control_transformer.py (hunk 2 above)
Comment (felixmaximilian) [reply src/rmind/models/control_transformer.py]: Done.

File: src/rmind/models/control_transformer.py (hunk 3 above)
Comment (felixmaximilian) [reply src/rmind/models/control_transformer.py]: This logic is gone, since we now use module.compile

File: src/rmind/callbacks/precision.py (hunk 1 above)
Comment (felixmaximilian) [reply src/rmind/callbacks/precision.py]: Went with option (b): added `rmind.utils.precision.auto_precision` and call it on `cfg.trainer.precision` right before `instantiate(cfg.trainer)` in `train.py`/`predict.py`. The Trainer is built with the resolved precision from the start — no more plugin surgery. Kept the auto-downgrade behavior (bf16-mixed → 16-mixed on Turing) since that was the original intent; logs a warning when it fires. Callback and its config entries are gone. d52390c

File: config/datamodule/yaak/train_debug.yaml (hunk 6 above)
Comment (felixmaximilian) [reply config/datamodule/yaak/train_debug.yaml]: Done. val.dataset now also pulls from /dataset/yaak/train_debug. 3d44c0d

File: config/datamodule/yaak/train_debug.yaml (hunk 7 above)
Comment (felixmaximilian) [reply config/datamodule/yaak/train_debug.yaml]: Reduced both batch sizes to 2. 3d44c0d

File: config/experiment/yaak/control_transformer/pretrain_debug.yaml (hunk 8 above)
Comment (felixmaximilian) [reply config/experiment/yaak/control_transformer/pretrain_debug.yaml]: Deleted the experiment config; moved its overrides inline into the train-debug justfile target as Hydra args (datamodule=yaak/train_debug + ++model.encoder.disable=true). 3d44c0d

File: flake.nix (hunk 4 above)
Comment (felixmaximilian) [reply flake.nix]: Switched to nix-managed compiler — stdenv.cc is now in the devshell packages, and CC/CXX export from there instead of /usr/bin. Added a comment explaining torch.compile (inductor) JIT-compiles fused kernels via the C++ toolchain. 3d44c0d

File: tests/test_compile_benchmark.py
Comment (felixmaximilian) [reply tests/test_compile_benchmark.py:1]: Would leave it outside of the scope of this PR.

**Comment by felixmaximilian** [comment]
PR is ready for final review

**Comment by egorchakov** [comment]
@codex review

File: README.md [hunk 10]
```diff
@@ -42,6 +42,18 @@ Foundation models for spatial intelligence.
 just train experiment=yaak/control_transformer/pretrain [...]
 ```
 
+Training uses `torch.compile` on the encoder by default (set in the model config). To disable it, pass `compile=false`.
```
[the remark this answers is not in the document]
Comment (felixmaximilian) [reply README.md]: README now reflects the new way of activating/deactivating the compilation

**Review by egorchakov — APPROVED** [review]
